File: src/CreateKude copy.ts
import { XMLParser } from 'fast-xml-parser';
import * as fs from 'fs';
import * as path from 'path';

// Using any type for pdfkit to avoid TypeScript compilation issues
const PDFDocument: any = require('pdfkit');

interface DocumentData {
  tipoDocumento: number;
  tipoDocumentoDescripcion: string;
  timbrado: string;
  establecimiento: string;
  punto: string;
  numero: string;
  serie: string | null;
}

interface JasperParameters {
  REPORT_LOCALE?: string;
  [key: string]: any;
}

interface ItemData {
  dCodInt?: string;
  dDesProSer?: string;
  dCantProSer?: string | number;
  dPUniProSer?: string | number;
  dTotBruOpeItem?: string | number;
}

type DocumentType = 1 | 2 | 3 | 4 | 5 | 6 | 7;

class CreateKude {
  private xmlParser: XMLParser;

  constructor() {
    this.xmlParser = new XMLParser({
      ignoreAttributes: false,
      attributeNamePrefix: '@_'
    });
  }

  public static async main(args: string[]): Promise<void> {
    const instance = new CreateKude();
    await instance.run(args);
  }

  private getTemplateName(tipoDocumento: DocumentType): string {
    switch (tipoDocumento) {
      case 1: return 'Factura.jasper';
      case 2: return 'FacturaImportacion.jasper';
      case 3: return 'FacturaExportacion.jasper';
      case 4: return 'AutoFactura.jasper';
      case 5: return 'NotaCredito.jasper';
      case 6: return 'NotaDebito.jasper';
      case 7: return 'NotaRemision.jasper';
      default: throw new Error(`Tipo de documento no soportado: ${tipoDocumento}`);
    }
  }

  private async run(args: string[]): Promise<void> {
    try {
      let xml: string | null = null;

      if (args.length > 0 && args[0] != null) {
        xml = args[0];
      } else {
        // Default path for development
        xml = "C:\\Users\\marco\\git\\tipscloud\\KUDE\\resources\\Extructura xml_DE.xml";
      }

      const dataFromXML = this.getTipoDocumentoFromXML(xml);
      const tipoDocumento = dataFromXML.tipoDocumento;

      let jasperPath: string | null = null;
      if (args.length > 1 && args[1] != null) {
        jasperPath = args[1];
        console.log("jasper path del par 1 vale " + jasperPath);

        let pathDestino: string | null = null;
        if (args.length > 2 && args[2] != null) {
          pathDestino = args[2];

          let parametros: Record<string, any> = {};
          if (args.length > 3 && args[3] != null) {
            try {
              parametros = JSON.parse(args[3]);
            } catch (error) {
              console.error("Error ignorado... No se pudo convertir " + args[3] + " a Objeto!");
            }
          }

          // Add locale parameter
          parametros.REPORT_LOCALE = 'es_PY';

          const templateName = this.getTemplateName(tipoDocumento as DocumentType);

          const fullJasperPath = path.join(jasperPath, templateName);
          console.log("----" + fullJasperPath);

          if (!fs.existsSync(fullJasperPath)) {
            throw new Error("Archivo " + fullJasperPath + " no encontrado.!");
          }

          try {
            // Generate PDF using Node.js PDF generation
            // Since JasperReports templates are binary, we'll create a basic PDF structure
            // In a full implementation, you might want to convert Jasper templates to a different format
            const pdfBuffer = await this.generatePDF(xml!, dataFromXML, parametros);

            const tipoDocumentoDescripcion = dataFromXML.tipoDocumentoDescripcion;
            const timbrado = dataFromXML.timbrado;
            const establecimiento = dataFromXML.establecimiento;
            const punto = dataFromXML.punto;
            const numero = dataFromXML.numero;
            const serie = dataFromXML.serie;

            const fileName = `${tipoDocumentoDescripcion}_${timbrado}-${establecimiento}-${punto}-${numero}${serie ? '-' + serie : ''}.pdf`;
            const fullPathDestino = path.join(pathDestino, fileName);

            fs.writeFileSync(fullPathDestino, pdfBuffer);
            console.log(`PDF generated successfully: ${fullPathDestino}`);
          } catch (error) {
            console.error('Error generating PDF:', error);
            throw error;
          }
        } else {
          throw new Error("Debe indicar el Path Destino para generación del PDF");
        }
      } else {
        throw new Error("Debe indicar el Path Origen de los archivos .jasper");
      }
    } catch (error) {
      console.error('Error:', error);
      process.exit(1);
    }
  }

  private generatePDF(xml: string, documentData: DocumentData, parametros: Record<string, any>): Promise<Buffer> {
    return new Promise((resolve, reject) => {
      try {
        const doc = new PDFDocument();
        const chunks: Buffer[] = [];

        doc.on('data', (chunk: Buffer) => chunks.push(chunk));
        doc.on('end', () => resolve(Buffer.concat(chunks)));
        doc.on('error', reject);

        // Parse XML to extract item data (similar to JasperReports data source)
        let xmlContent: string;
        if (xml.startsWith('<?xml')) {
          xmlContent = xml;
        } else {
          xmlContent = fs.readFileSync(xml, 'utf8');
        }

        const parsedXML = this.xmlParser.parse(xmlContent);

        // Header
        doc.fontSize(18).text('KUDE - Factura Electrónica', { align: 'center' });
        doc.moveDown();

        // Document info
        doc.fontSize(12)
          .text(`Tipo Documento: ${documentData.tipoDocumentoDescripcion}`)
          .text(`Timbrado: ${documentData.timbrado}`)
          .text(`Establecimiento: ${documentData.establecimiento}`)
          .text(`Punto: ${documentData.punto}`)
          .text(`Número: ${documentData.numero}`);
        if (documentData.serie) {
          doc.text(`Serie: ${documentData.serie}`);
        }
        doc.moveDown();

        // Try to extract and display items from XML
        try {
          const rDE = parsedXML.rDE;
          const DE = rDE.DE;
          const gCamItems = DE.gCamItem;

          if (gCamItems) {
            doc.fontSize(14).text('Items:', { underline: true });
            doc.moveDown();

            const items = Array.isArray(gCamItems) ? gCamItems : [gCamItems];

            items.forEach((item: any, index: number) => {
              doc.fontSize(10)
                .text(`Item ${index + 1}:`)
                .text(`  Código: ${item.dCodInt || 'N/A'}`)
                .text(`  Descripción: ${item.dDesProSer || 'N/A'}`)
                .text(`  Cantidad: ${item.dCantProSer || 'N/A'}`)
                .text(`  Precio: ${item.dPUniProSer || 'N/A'}`);
              if (item.dTotBruOpeItem) {
                doc.text(`  Total: ${item.dTotBruOpeItem}`);
              }
              doc.moveDown(0.5);
            });
          }
        } catch (error) {
          doc.fontSize(10).text('No se pudieron extraer los items del XML');
          console.warn('Error extracting items:', error);
        }

        // Add parameters if any
        if (Object.keys(parametros).length > 0) {
          doc.moveDown();
          doc.fontSize(12).text('Parámetros:', { underline: true });
          doc.fontSize(8).text(JSON.stringify(parametros, null, 2));
        }

        doc.moveDown();
        doc.fontSize(8).text('Generado con Node.js/TypeScript - Reemplazo de JasperReports', { align: 'center' });

        doc.end();
      } catch (error) {
        reject(error);
      }
    });
  }

  private getTipoDocumentoFromXML(archivoDEXML: string): DocumentData {
    try {
      let xmlContent: string;

      if (archivoDEXML.startsWith('<?xml')) {
        xmlContent = archivoDEXML;
      } else {
        xmlContent = fs.readFileSync(archivoDEXML, 'utf8');
      }

      const parsedXML = this.xmlParser.parse(xmlContent);

      // Navigate through the XML structure as in the Java version
      const rDE = parsedXML.rDE;
      const DE = rDE.DE;
      const gTimb = DE.gTimb;

      const tipoDocumento = parseInt(DE.iTiDE);
      const tipoDocumentoDescripcion = DE.dDesTiDE;
      const timbrado = DE.dNumTim;
      const establecimiento = DE.dEst;
      const punto = DE.dPunExp;
      const numero = DE.dNumDoc;
      const serie = DE.dSerieNum || null;

      return {
        tipoDocumento,
        tipoDocumentoDescripcion,
        timbrado,
        establecimiento,
        punto,
        numero,
        serie
      };
    } catch (error) {
      throw new Error(`Error parsing XML: ${error.message}`);
    }
  }
}

// For command line usage
if (require.main === module) {
  CreateKude.main(process.argv.slice(2)).catch(console.error);
}

export default CreateKude;

File: src/CreateKude copy.java
 import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import net.sf.jasperreports.engine.JasperExportManager;
import net.sf.jasperreports.engine.JasperFillManager;
import net.sf.jasperreports.engine.JasperPrint;
import net.sf.jasperreports.engine.data.JRXmlDataSource;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

public class CreateKude {
   public static void main(String[] args) throws Exception {
      Logger.getRootLogger().setLevel(Level.ERROR);
      String xml = null;
      if (args.length > 0 && args[0] != null) {
         xml = args[0];
      } else {
         xml = "C:\\Users\\marco\\git\\tipscloud\\KUDE\\resources\\Extructura xml_DE.xml";
      }

      Map<String, Object> dataFromXML = getTipoDocumentoFromXML(xml);
      int tipoDocumento = (Integer)dataFromXML.get("tipoDocumento");
      String jasperPath = null;
      if (args.length > 1 && args[1] != null) {
         jasperPath = args[1];
         System.out.println("jasper path del par 1 vale " + jasperPath);
         String pathDestino = null;
         if (args.length > 2 && args[2] != null) {
            pathDestino = args[2];
            Object parametros = new HashMap();
            Gson inputStream;
            if (args.length > 3 && args[3] != null) {
               inputStream = (new GsonBuilder()).setDateFormat("yyyy-MM-dd'T'HH:mm:ss").create();

               try {
                  parametros = (Map)inputStream.fromJson(args[3], parametros.getClass());
               } catch (Exception var18) {
                  System.err.println("Error ignorado... No se pudo convertir " + args[3] + " a Objeto!");
               }
            }

            if (tipoDocumento == 1) {
               jasperPath = jasperPath + "Factura.jasper";
            }

            if (tipoDocumento == 2) {
               jasperPath = jasperPath + "FacturaImportacion.jasper";
            }

            if (tipoDocumento == 3) {
               jasperPath = jasperPath + "FacturaExportacion.jasper";
            }

            if (tipoDocumento == 4) {
               jasperPath = jasperPath + "AutoFactura.jasper";
            }

            if (tipoDocumento == 5) {
               jasperPath = jasperPath + "NotaCredito.jasper";
            }

            if (tipoDocumento == 6) {
               jasperPath = jasperPath + "NotaDebito.jasper";
            }

            if (tipoDocumento == 7) {
               jasperPath = jasperPath + "NotaRemision.jasper";
            }

            System.out.println("----" + jasperPath);
            if (!(new File(jasperPath)).exists()) {
               throw new Exception("Archivo " + jasperPath + " no encontrado.!");
            } else {
               try {
                  inputStream = null;
                  Object inputStream;
                  if (xml.startsWith("<?xml")) {
                     inputStream = new ByteArrayInputStream(xml.getBytes());
                  } else {
                     inputStream = new FileInputStream(xml);
                  }

                  JRXmlDataSource dataSource = new JRXmlDataSource((InputStream)inputStream, "/rDE/DE/gDtipDE/gCamItem");
                  Locale locale = new Locale("es", "PY");
                  ((Map)parametros).put("REPORT_LOCALE", locale);
                  JasperPrint jprint = JasperFillManager.fillReport(jasperPath, (Map)parametros, dataSource);
                  String tipoDocumentoDescripcion = (String)dataFromXML.get("tipoDocumentoDescripcion");
                  String timbrado = (String)dataFromXML.get("timbrado");
                  String establecimiento = (String)dataFromXML.get("establecimiento");
                  String punto = (String)dataFromXML.get("punto");
                  String numero = (String)dataFromXML.get("numero");
                  String serie = (String)dataFromXML.get("serie");
                  pathDestino = pathDestino + tipoDocumentoDescripcion + "_" + timbrado + "-" + establecimiento + "-" + punto + "-" + numero + (serie != null ? "-" + serie : "") + ".pdf";
                  JasperExportManager.exportReportToPdfFile(jprint, pathDestino);
               } catch (Exception var17) {
                  var17.printStackTrace();
               }

            }
         } else {
            throw new Exception("Debe indicar el Path Destino para generación del PDF");
         }
      } else {
         throw new Exception("Debe indicar el Path Origen de los archivos .jasper");
      }
   }

   public static Map<String, Object> getTipoDocumentoFromXML(String archivoDEXML) throws Exception {
      Map<String, Object> dataFromXML = new HashMap();
      File file = new File(archivoDEXML);
      DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
      DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
      Document document = null;
      if (archivoDEXML.startsWith("<?xml")) {
         InputStream inputStream = new ByteArrayInputStream(archivoDEXML.getBytes());
         document = documentBuilder.parse(inputStream);
      } else {
         document = documentBuilder.parse(file);
      }

      Node student = document.getElementsByTagName("rDE").item(0);
      Element dataFileElement = (Element)student;
      Node dataFile = dataFileElement.getElementsByTagName("DE").item(0);
      Element locationElement = (Element)dataFile;
      Node location = locationElement.getElementsByTagName("gTimb").item(0);
      Element iTipoDEElement = (Element)dataFile;
      Node iTiDE = iTipoDEElement.getElementsByTagName("iTiDE").item(0);
      Integer tipoDocumento = Integer.valueOf(iTiDE.getTextContent());
      Node dDesTiDE = iTipoDEElement.getElementsByTagName("dDesTiDE").item(0);
      String tipoDocumentoDescripcion = dDesTiDE.getTextContent();
      Node dNumTim = iTipoDEElement.getElementsByTagName("dNumTim").item(0);
      String timbrado = dNumTim.getTextContent();
      Node dEst = iTipoDEElement.getElementsByTagName("dEst").item(0);
      String establecimiento = dEst.getTextContent();
      Node dPunExp = iTipoDEElement.getElementsByTagName("dPunExp").item(0);
      String punto = dPunExp.getTextContent();
      Node dNumDoc = iTipoDEElement.getElementsByTagName("dNumDoc").item(0);
      String numero = dNumDoc.getTextContent();
      Node dSerieNum = iTipoDEElement.getElementsByTagName("dSerieNum").item(0);
      String serie = null;
      if (dSerieNum != null) {
         dSerieNum.getTextContent();
      }

      dataFromXML.put("tipoDocumento", tipoDocumento);
      dataFromXML.put("tipoDocumentoDescripcion", tipoDocumentoDescripcion);
      dataFromXML.put("timbrado", timbrado);
      dataFromXML.put("establecimiento", establecimiento);
      dataFromXML.put("punto", punto);
      dataFromXML.put("numero", numero);
      dataFromXML.put("serie", serie);
      return dataFromXML;
   }
}
    